docs/flatpak-devel.rst

```diff
@@ -0,0 +1,395 @@
+
+The CI setup will vary based on the code host and the CI service being
+used. A basic workflow for exporting to Gitlab pages is provided in
+:doc:`hosting-a-repository` which can be used as a starting point.
+
+`GNOME <https://gitlab.gnome.org/GNOME/Initiatives/-/wikis/DevOps-with-Flatpak#basic-ci>`_
+and `KDE <https://develop.kde.org/docs/packaging/flatpak/publishing/#publishing-to-kdes-nightly-repositories>`_
+also has their own respective CI setup for doing this, which can be
+also be used for inspiration.
+
+`Flatpak Github Actions <https://github.com/flathub-infra/flatpak-github-actions>`_
+can be used for GitHub.
+
+Running tests
+-------------
+
+Flatpak Builder can also run tests with ``run-tests, test-rule, test-commands, test-args``.
+
+``run-tests: true`` in a module will run tests for that module after
+installing.
+
+``test-rule`` is the default target to build when running tests. For
+``meson`` an ``cmake-ninja`` buildsystems it defaults to ``test``,
+otherwise ``check`` for ``make``. Setting it to an empty string like
+``test-rule: ''`` will disable the target.
+
+``test-commands`` is an array of additional commands that will be run
+during tests.
+
+.. code-block:: yaml
+
+  run-tests: true
+  test-rule: ''
+  test-commands:
+     - xvfb-run tests/test_foo
+
+``test-args`` is used to provide finish-args for tests. These do not
+affect the normal installation.
+
+.. code-block:: yaml
+
+  build-options:
+      test-args:
+          - "--socket=x11"
+          - "--share=network"
+
+Parallel nigthly and stable applications
+----------------------------------------
+
+In general stable and nightly versions or any two parallel branches
+of an application should have different application IDs. This is
+mandatory if the application, uses its own well known Dbus name as the
+dbus names need to be distinct.
+
+A standard way is to use the ``.Devel`` suffix to the original Flatpak ID.
+The internal application ID, appstream ID, launchable, icon tags,
+icon files, desktop files, Dbus service names etc. must follow the new
+Flatpak ID for everything to work seamlessly for the user.
+
+This can be integrated with buildsystem `profiles`, so simply building
+with ``-Dprofile=Devel`` for example will build the Nightly application
+and the default will build the stable appliction.
+
+Buildsystem handling
+^^^^^^^^^^^^^^^^^^^^
+
+This is an example for the Meson buildsystem.
+
+The ``src`` subdirectory will contain the main source code of the
+application and ``data`` subdirectory will have application metadata
+such as desktop file, metainfo file, icons and the ``po`` subdirectory
+will have translation files.
+
+The root ``meson.build`` parts should be something like this:
+
+.. code-block::
+
+  i18n = import('i18n')
+
+  profile = get_option('profile')
+  if profile == 'development'
+    find_program('git')
+    rev_txt = run_command('git','rev-parse','--short','HEAD').stdout().strip()
```

Comment:
AFAICT this information is never used?

There is also the `vcs_tag()` function that overlaps with this, depending on needs.

Reply:
removed.